### PR TITLE
range: handle overflow errors when calculating ranges

### LINKED ISSUE #17253: select statement returns constant overflow
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
tidb> create table t6 (a bigint);
Query OK, 0 rows affected (0.00 sec)
tidb> insert into t6 values(9223372036854775807);
Query OK, 1 row affected (0.00 sec)
tidb> alter table t6 add unique key idx(a);
Query OK, 0 rows affected (0.04 sec)
mysql> select * from t6 where a != 9223372036854775808;
ERROR 1690 (22003): constant 9223372036854775808 overflows bigint
```

### 2. What did you expect to see? (Required)
```
mysql> select * from t6 where a != 9223372036854775808;
+---------------------+
| a                   |
+---------------------+
| 9223372036854775807 |
+---------------------+
1 row in set (0.00 sec)
```

### 3. Affected version (Required)
master

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #17253 

Problem Summary: when calculating ranges for expressions like `bigint_column < overflow_value`, we should convert it to `bigint_column <= max_bigint_value` instead of returning an overflow error.

### What is changed and how it works?
Handle this error when calculating ranges.

### Check List 

Tests 

- Unit test

### Release note 
- Handle overflow errors when calculating ranges

### CODE REVIEW

File: util/ranger/ranger.go [hunk 1]
```diff
@@ -85,22 +91,51 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
 	return ranges, nil
 }
 
-func convertPoint(sc *stmtctx.StatementContext, point point, tp *types.FieldType) (point, error) {
+func convertPoint(sc *stmtctx.StatementContext, point point, isLeft bool, tp *types.FieldType) (_ point, invalid bool, _ error) {
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:94]: What does `isLeft` mean?

Comment (XuHuaiyu) [inline util/ranger/ranger.go:94]: `return valid` maybe easier to understand

File: util/ranger/ranger.go [hunk 2]
```diff
@@ -85,22 +91,51 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
 	casted, err := point.value.ConvertTo(sc, tp)
 	if err != nil {
-		return point, errors.Trace(err)
+		if types.ErrOverflow.Equal(err) && tp.EvalType() == types.ETInt {
+			lowBound, upBound := types.GetMinValue(tp), types.GetMaxValue(tp)
+			if r, err := casted.CompareDatum(sc, &lowBound); err != nil {
+				return point, false, errors.Trace(err)
+			} else if r == 0 { // point < lowBound
+				if isLeft {
+					// (-inf < lowBound => [lowBound
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:107]: what does this comment mean..

File: util/ranger/ranger.go [hunk 3]
```diff
@@ -85,22 +91,51 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
+func convertPoint(sc *stmtctx.StatementContext, point point, isLeft bool, tp *types.FieldType) (_ point, invalid bool, _ error) {
 	switch point.value.Kind() {
 	case types.KindMaxValue, types.KindMinNotNull:
-		return point, nil
+		return point, false, nil
 	}
 	casted, err := point.value.ConvertTo(sc, tp)
 	if err != nil {
-		return point, errors.Trace(err)
+		if types.ErrOverflow.Equal(err) && tp.EvalType() == types.ETInt {
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:101]: if !types.ErrOverflow.Equal(err) || tp.EvalType() != types.ETInt{
    return
}

File: util/ranger/ranger.go [hunk 4]
```diff
@@ -85,22 +91,51 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
+			lowBound, upBound := types.GetMinValue(tp), types.GetMaxValue(tp)
+			if r, err := casted.CompareDatum(sc, &lowBound); err != nil {
+				return point, false, errors.Trace(err)
+			} else if r == 0 { // point < lowBound
+				if isLeft {
+					// (-inf < lowBound => [lowBound
+					point.value = lowBound
+					point.excl = false
+					return point, false, nil
+				} else {
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:111]: useless `else`

**Comment by qw4990** [comment]
I add some comments and update these lines of code, PTAL @XuHuaiyu

File: util/ranger/ranger.go [hunk 5]
```diff
@@ -91,43 +91,54 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
 	return ranges, nil
 }
 
+// convertPoint converts this point according to the specified type.
+// isLeft is used to specify if this point is the left point in a range.
+// for any range (leftPoint, rightPoint),
+//		if the leftPoint is larger than the upperBound of this type, or
+//		the rightPoint is less than the lowerBound of this type,
+//		we think it's an invalid range and return with invalid=true.
+//	Here are some examples with the bound range of this type is [10, 20]:
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:100]: If we use type `tinyint` here to make this example, it'll be easier to understand.

File: util/ranger/ranger.go [hunk 6]
```diff
@@ -91,43 +91,54 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
-				}
+		if !types.ErrOverflow.Equal(err) || tp.EvalType() != types.ETInt {
+			return point, false, errors.Trace(err)
+		}
+		// handle overflow errors for int especially
+		lowBound, upBound := types.GetMinValue(tp), types.GetMaxValue(tp)
+		var r int
+		if r, err = casted.CompareDatum(sc, &lowBound); err != nil {
+			return point, false, errors.Trace(err)
+		} else if r == 0 { // point < lowBound
```
Comment (XuHuaiyu) [inline util/ranger/ranger.go:123]: 1. It seems that `r == 0` indicates `point == lowBound`?
2. What's the result of [0.1, 1.1] ?

File: util/ranger/ranger.go (hunk 5 above)
Comment (qw4990) [reply util/ranger/ranger.go]: Updated, PTAL.

File: util/ranger/ranger.go [hunk 7]
```diff
@@ -85,22 +91,62 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
-		return point, errors.Trace(err)
+		if !types.ErrOverflow.Equal(err) || tp.EvalType() != types.ETInt {
+			return point, false, errors.Trace(err)
+		}
+		// handle overflow errors for int especially
+		lowBound, upBound := types.GetMinValue(tp), types.GetMaxValue(tp)
+		var r int
+		if r, err = casted.CompareDatum(sc, &lowBound); err != nil {
+			return point, false, errors.Trace(err)
+		} else if r == 0 { // point < lowBound
```
Comment (winoros) [inline util/ranger/ranger.go:123]: We can comment more here to help understand codes.

File: util/ranger/ranger.go [hunk 8]
```diff
@@ -85,22 +91,62 @@ func points2Ranges(sc *stmtctx.StatementContext, rangePoints []point, tp *types.
...
+			if isLeft { // example 1
+				point.value = lowBound
+				point.excl = false
+			} else { // example 2
+				invalid = true
+			}
+		}
+		if r, err := casted.CompareDatum(sc, &upBound); err != nil {
+			return point, false, errors.Trace(err)
+		} else if r == 0 { // point > upBound
```
Comment (winoros) [inline util/ranger/ranger.go:133]: Same.

**Comment by qw4990** [comment]
Since this way to fix #17253 meets some unsolved problems, it is closed now, and we will use another PR to fix it soon.

**Pull request closed without merging**